code/modules/telesci/gps.dm

```diff
@@ -13,69 +11,145 @@ var/global/list/gps_by_type = list()
 
 /obj/item/device/gps/Destroy()
-	GPS_list -= src
-	var/list/typelist = gps_by_type["[type]"]
-	LAZYREMOVE(typelist, src)
+	GPS_list -= GPS_list[gpstag]
+	moved_event.unregister(src, src)
+	if(held_by)
+		moved_event.unregister(held_by, src)
+		held_by = null
 	return ..()
 
+/obj/item/device/gps/pickup(var/mob/user)
+	..()
+	held_by = user
+	moved_event.register(user, src, /obj/item/device/gps/proc/update_position)
+
+/obj/item/device/gps/dropped(var/mob/user)
+	..()
+	held_by = null
+	moved_event.unregister(user, src)
+
 /obj/item/device/gps/emp_act(severity)
 	emped = 1
 	cut_overlay("working")
 	add_overlay("emp")
 	addtimer(CALLBACK(src, .proc/post_emp), 300)
+	update_position()
 
 /obj/item/device/gps/proc/post_emp()
 	emped = 0
 	cut_overlay("emp")
 	add_overlay("working")
+	update_position()
 
-/obj/item/device/gps/attack_self(mob/user)
+/obj/item/device/gps/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
+	if(!data)
+		. = data = list()
+	VUEUI_SET_CHECK(data["own_tag"], gpstag, ., data)
+	
+	var/list/tracking_list = list()
+	for(var/tracking_tag in sortList(tracking))
+		if(!GPS_list[tracking_tag]) // Another GPS device has changed its tag or been destroyed
+			tracking -= tracking_tag
+			continue
 
-	var/obj/item/device/gps/t = ""
-	var/gps_window_height = 110 + GPS_list.len * 20 // Variable window height, depending on how many GPS units there are to show
-	if(emped)
-		t += "ERROR"
-	else
-		t += "<BR><A href='?src=\ref[src];tag=1'>Set Tag</A> "
-		t += "<BR>Tag: [gpstag]"
-		if(locked_location && locked_location.loc)
-			t += "<BR>Bluespace coordinates saved: [locked_location.loc]"
-			gps_window_height += 20
-
-		for(var/obj/item/device/gps/G in GPS_list)
-			var/turf/pos = get_turf(G)
-			var/area/gps_area = get_area(G)
-			var/tracked_gpstag = G.gpstag
-			if(G.emped == 1 || !pos)
-				t += "<BR>[tracked_gpstag]: ERROR"
-			else
-				t += "<BR>[tracked_gpstag]: [format_text(gps_area.name)] ([pos.x], [pos.y], [pos.z])"
+		if(AreConnectedZLevels(src.loc.z, GPS_list[tracking_tag]["pos_z"]))
+			tracking_list[tracking_tag] = (GPS_list[tracking_tag])
+
+	data["tracking_list"] = tracking_list
+
+/obj/item/device/gps/attack_self(mob/user)
+	if(!emped)
+		ui_interact(user)
 
-	var/datum/browser/popup = new(user, "GPS", name, 360, min(gps_window_height, 800))
-	popup.set_content(t)
-	popup.set_title_image(user.browse_rsc_icon(src.icon, src.icon_state))
-	popup.open()
+/obj/item/device/gps/ui_interact(mob/user)
+	var/datum/vueui/ui = SSvueui.get_open_ui(user, src)
+	if (!ui)
+		ui = new(user, src, "devices-gps-gps", 460, 600, capitalize(src.name))
+		ui.auto_update_content = TRUE
+	ui.open()
 
 /obj/item/device/gps/Topic(href, href_list)
 	..()
+
 	if(href_list["tag"] )
-		var/a = input("Please enter desired tag.", name, gpstag) as text
-		a = uppertext(copytext(sanitize(a), 1, 5))
+		var/set_tag = input("Please enter desired tag.", name, gpstag) as text
+		set_tag = uppertext(copytext(sanitize(set_tag), 1, 8))
```

Comment:
Can we get by without `input()`? Use text box at UI for editing tag.

Reply:
Text box'd it. Thank you.